tools/drivewatch.go

```diff
@@ -575,6 +579,122 @@ func (dw *driveWatch) needRingUpdate(rd ringData) bool {
+		return err
+	}
+	defer rows.Close()
+	missingPartitions := map[int64]string{}
+	for rows.Next() {
+		var p int64
+		var a, c, o string
+		if err := rows.Scan(&p, &a, &c, &o); err == nil {
+			missingPartitions[p] = fmt.Sprintf("%s/%s/%s",
+				common.Urlencode(a), common.Urlencode(c), common.Urlencode(o))
+		} else {
+			dw.logger.Error("SELECT for checkMissingDispersion", zap.Error(err))
+		}
+	}
+	longMissingPartitions := []int64{}
+
+	for p := range missingPartitions {
+
+		rows, err := tx.Query(`
+			SELECT drd.partition
+			FROM dispersion_report dr LEFT JOIN dispersion_report_detail drd
+			ON dr.id = drd.dispersion_report_id AND drd.partition = ?
+			WHERE dr.policy = ? AND dr.rtype = ? AND dr.create_date > ?
+			`, p, policy, "object", time.Now().Add(-dw.rescueMissingPartAge))
+
+		if err != nil {
+			dw.logger.Error("SELECT for reports checkMissingDispersionObjects", zap.Error(err))
+			return err
+		}
+		defer rows.Close()
+		isReportWithPart := false
+		for rows.Next() {
+			var p sql.NullInt64
+			if err := rows.Scan(&p); err == nil {
+				if p.Int64 == 0 {
+					// if part not in report, all copies were found
+					isReportWithPart = true
+				}
+			} else {
+				dw.logger.Error("SELECT for reports checkMissingDispersionObjects scan", zap.Error(err))
+				continue
+			}
+		}
+		if !isReportWithPart {
+			longMissingPartitions = append(longMissingPartitions, p)
+		}
+	}
+	dw.logger.Info("Dispersion heal start", zap.Int("numPartitions", len(longMissingPartitions)))
+	for i := len(longMissingPartitions) - 1; i > 0; i-- { // shuffle
+		j := rand.Intn(i + 1)
+		longMissingPartitions[j], longMissingPartitions[i] = longMissingPartitions[i], longMissingPartitions[j]
+	}
+	for _, p := range longMissingPartitions {
+		select {
+		case <-cancel:
+			break
+		default:
+		}
+		nodes := rd.r.GetNodes(uint64(p))
+		goodNodes := []*ring.Device{}
+		notFoundNodes := []*ring.Device{}
+		for _, device := range nodes {
+			url := fmt.Sprintf("http://%s:%d/%s/%d/%s", device.Ip, device.Port, device.Device, p, missingPartitions[p])
+			req, _ := http.NewRequest("HEAD", url, nil)
+			resp, err := dw.client.Do(req)
+
+			if err == nil && resp.StatusCode/100 == 2 {
+				goodNodes = append(goodNodes, device)
+			} else if resp != nil && resp.StatusCode == 404 {
+				notFoundNodes = append(notFoundNodes, device)
+			} else {
+				dw.logger.Error("Dispersion heal invalid response.", zap.String("url", url), zap.Error(err))
+			}
+			if resp != nil {
+				resp.Body.Close()
+			}
+		}
+		if len(goodNodes) == 0 {
+			dw.logger.Error("Dispersion heal found no good nodes.", zap.String("partPath", missingPartitions[p]))
+			continue
+		}
+		if len(notFoundNodes) == 0 {
+			continue
+		}
+		goodNode := goodNodes[rand.Intn(len(goodNodes))]
+		resultChan := make(chan string, 1)
+		go rescueLonelyPartition(int64(policy), uint64(p), goodNode, notFoundNodes, rd.r.GetMoreNodes(uint64(p)), resultChan, dw.client)
+		res := <-resultChan
```

Comment:
Did you intend for this to be single concurrency? Or did you plan to iterate on the code and make it concurrent later?

Reply:
wanted single conc- for now anyway